### PR TITLE
Allow caching the isolated build environment pip

### PR DESCRIPTION
**Pull request by gaborbernat**
- fix setuptools is already provisioned for a fresh isolated environment

Related https://github.com/pypa/pep517/pull/97

Signed-off-by: Bernát Gábor <bgabor8@bloomberg.net>

### CODE REVIEW

File: src/build/__main__.py [hunk 1]
```diff
@@ -125,65 +130,88 @@ def main_parser():  # type: () -> argparse.ArgumentParser
...
+        dest='cache',
+        help='cache isolation environment(s) in between runs',
+        default=default_isolation.cache,
+    )
+    group.add_argument(
+        '--reset-cache',
+        dest='reset_cache',
+        help='clear out the tools cache',
+        action='store_true',
+        default=default_isolation.reset_cache,
```
Comment (FFY00) [inline src/build/__main__.py:170]: I don't like exposing this. This leaks implementation details all over.

File: src/build/__main__.py [hunk 2]
```diff
@@ -96,7 +96,12 @@ def main_parser():  # type: () -> argparse.ArgumentParser
     """
     cwd = os.getcwd()
     out = os.path.join(cwd, 'dist')
-    parser = argparse.ArgumentParser()
+
+    class HelpFormatter(argparse.ArgumentDefaultsHelpFormatter):
+        def __init__(self, prog):
+            super(HelpFormatter, self).__init__(prog, max_help_position=32, width=240)
+
+    parser = argparse.ArgumentParser(formatter_class=HelpFormatter)
```
Comment (FFY00) [inline src/build/__main__.py:104]: This is kinda unrelated, can it be in a separate PR?

File: src/build/env.py [hunk 3]
```diff
@@ -215,27 +259,74 @@ def __enter__(self):  # type: () -> IsolatedEnvironment
...
+    def _provision_pip(self):
+        if self._isolation.ensure_pip:
+            subprocess.check_call([self._executable, '-Im', 'ensurepip', '--upgrade', '--default-pip'])
+            # we don't want the default setuptools
+            subprocess.check_call([self._executable, '-m', 'pip', 'uninstall', 'setuptools', '-y'])
+        else:
+            # version qualify per python minor, so that we don't break when pip drops older python versions
+            folder = os.path.expanduser(os.path.expandvars(self._isolation.cache))
+            if self._isolation.reset_cache and os.path.exists(folder):
+                print('Delete {}'.format(folder))
```
Comment (FFY00) [inline src/build/env.py:301]: Remove this print. We could add a logging system if needed, but we should not be using print IMO.

**Review by FFY00 — CHANGES_REQUESTED** [review]
Okay, I am not so happy with this implementation.

First, I disklike us caching the build environments. It is something I think can easily go wrong, when we reuse them we need to make sure that *only* the expected requirements are present.

For now, I want to put a hold on caching the build environments. I am not even sure if it is something I want to have here. It's more public exports that we need to maintain. IMO right now we should only be caching the pip environment. The requirement wheels will be cached by pip, so environment creation shouldn't really that much time.

Can you remove `--ensurepip` and `--cache`? And rename `--reset-cache` to `--reset-tool-cache`, which would remove the cached pip.

File: src/build/env.py [hunk 4]
```diff
@@ -6,16 +6,50 @@
...
 _HAS_SYMLINK = None  # type: Optional[bool]
 
 
+DARWIN_CACHE = '~/Library/Caches/pypa/build'
+WINDOWS_CACHE = '~\\AppData\\Local\\pypa\\build\\Cache'
+LINUX_CACHE = '~/.cache/pypa/build'
+
+
+def get_shared_folder():
+    return DARWIN_CACHE if sys.platform == 'darwin' else WINDOWS_CACHE if sys.platform == 'win32' else LINUX_CACHE
```
[the remark this answers is not in the document]
Comment (gaborbernat) [reply src/build/env.py:25]: I mean ideally we'd just depend on appdirs as all other tools do, but you were not that keen on new dependencies.

File: src/build/env.py (hunk 3 above)
Comment (gaborbernat) [reply src/build/env.py:301]: I mean the most popular backend setuptools already uses print, and we don't interrupt that. So there's already tons of print in the code. So not sure I follow on your worry.

**Review by gaborbernat — COMMENTED** [review]
> First, I disklike us caching the build environments.

Well, there's no build environment caching here. Only the ensurepip is cached, so not sure what to make of this comment.

It's more public exports that we need to maintain.

I don't follow what you mean here.

> The requirement wheels will be cached by pip, so environment creation shouldn't really be that much time.

You're being overly optimistic. In the spirit of measure, instead of blindly trust:

```
Done isolated build environment at /var/folders/jh/kx1127wj32l8c3s97131rhmr0000gp/T/build-env-g8dy3sdj  within 0.01006
Looking in indexes: http://localhost:3141/bernat/bb/+simple
Collecting wheel>=0.30
  Downloading http://localhost:3141/bernat/bb/%2Bf/497/add53525d16c1/wheel-0.35.1-py2.py3-none-any.whl (33 kB)
Collecting setuptools>=44
  Downloading http://localhost:3141/bernat/bb/%2Bf/c77/b3920663a435c/setuptools-50.3.0-py3-none-any.whl (785 kB)
     |████████████████████████████████| 785 kB 71.1 MB/s
Collecting setuptools_scm[toml]>=3.4
  Downloading http://localhost:3141/bernat/bb/%2Bf/692/58e2eeba5f7ce/setuptools_scm-4.1.2-py2.py3-none-any.whl (27 kB)
Collecting toml; extra == "toml"
  Downloading http://localhost:3141/bernat/bb/%2Bf/bda/89d5935c2eac5/toml-0.10.1-py2.py3-none-any.whl (19 kB)
Installing collected packages: wheel, setuptools, toml, setuptools-scm
Successfully installed setuptools-50.3.0 setuptools-scm-4.1.2 toml-0.10.1 wheel-0.35.1
Done provision build dependencies wheel >= 0.30 setuptools >= 44 setuptools_scm[toml]>=3.4 within 2.521901
```

For reference ensurepip also takes around 2.5 second without the caching introduced here. And this is on macOs with SSD, where performance is better. Will be much worse on Windows with HDD.

> Can you remove --ensurepip and --cache? And rename --reset-cache to --reset-tool-cache, which would remove the cached pip.

I'm putting here my hat of experience with maintaining a such system already for virtualenv. People will need ``ensurepip`` as escape hatchet when for whatever reason the cache system doesn't work for them (e.g. permissioning to our choosen home folder or CI caches that persist only some part of the system). We can remove them but that'd be short sighted because we'd get these bug requests. These flags are there for advanced use cases, and should not be touched by default.

File: src/build/env.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (gaborbernat) [reply src/build/env.py:25]: > I'd respect `XDG_CACHE_HOME` on Linux, if set, otherwise use `~/.cache` for it.

This proposal is very UNIX friendly but you're alleniating Windows/macOs users. 😢 We can't be platform leaning in a PyPa tool.

[the remark this answers is not in the document]
Comment (FFY00) [reply src/build/env.py:25]: Sorry, I wasn't clear. What I was trying to say is that we should have a fallback, instead of hardcoding for all platforms.

I additionally suggested to respect `XDG_CACHE_HOME` on Linux, in the hardcoded path.

File: src/build/env.py [hunk 5]
```diff
@@ -6,16 +6,50 @@
 import sysconfig
 import tempfile
 import types
-
-from typing import Dict, Iterable, List, Optional, Sequence, Type
-
+from contextlib import contextmanager
```
[the remark this answers is not in the document]
Comment (gaborbernat) [reply src/build/env.py:9]: > It requires me to remember where `contextmanager` comes from

IMHO any decent editor will point out if you care about the source in a second. Furthermore, IMHO popular stdlib elements should be obvious enough mildly familiar with python that they make the code hard to read for anyone else... 🤷 I guess we'll need to agree to disagree on this.

File: src/build/__main__.py (hunk 1 above)
Comment (FFY00) [reply src/build/__main__.py:170]: I think we should only have `--reset-tool-cache` to remove up the pip environment. We can also have a `PYTHON_BUILD_NO_CACHE` environment variable if that's necessary, but I doubt it. The argument to use an env var instead of a CLI option would be that if you are in one of those special cases where you don't want pip to be cached, you probably want to do it for all calls, for eg. in a CI, you would only set the env var one time.

File: src/build/env.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (FFY00) [reply src/build/env.py:9]: I cannot do this easily in either my open source build of vscode, or in vim w/ jedi. I don't think it's that universal.

**Comment by gaborbernat** [comment]
sad walk
